### ISSUE TITLE
undefined symbol: PyByteArray_Type

### ISSUE THREAD

**Issue by outeleiloa** [vim/vim#3629]
When issuing the following command:
```
$ vim --cmd 'py3 import struct'
```

The following error is thrown:
```
Error detected while processing pre-vimrc command line:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.7/struct.py", line 13, in <module>
    from _struct import *
ImportError: /usr/local/lib/python3.7/lib-dynload/_struct.cpython-37m-x86_64-linux-gnu.so: undefined symbol: PyByteArray_Type
Press ENTER or type command to continue
```

The error only occurs when running the internal vim Python 3 interpreter.

vim was built from source and configured with the following:
```
./configure --enable-python3interp=yes --with-python3-command=python3 --enable-fail-if-missing
```

vim version command output:
```
$ vim --version
VIM - Vi IMproved 8.1 (2018 May 18, compiled Nov 21 2018 21:36:44)
Included patches: 1-535
...
Huge version without GUI.  Features included (+) or not (-):
+acl               +extra_search      +mouse_netterm     +tag_old_static
+arabic            +farsi             +mouse_sgr         -tag_any_white
+autocmd           +file_in_path      -mouse_sysmouse    -tcl
+autochdir         +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
-balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
-browse            +fork()            -mzscheme          +textobjects
++builtin_terms    +gettext           +netbeans_intg     +timers
+byte_offset       -hangul_input      +num64             +title
+channel           +iconv             +packages          -toolbar
+cindent           +insert_expand     +path_extra        +user_commands
-clientserver      +job               -perl              +vartabs
-clipboard         +jumplist          +persistent_undo   +vertsplit
+cmdline_compl     +keymap            +postscript        +virtualedit
+cmdline_hist      +lambda            +printer           +visual
+cmdline_info      +langmap           +profile           +visualextra
+comments          +libcall           -python            +viminfo
+conceal           +linebreak         +python3           +vreplace
+cryptv            +lispindent        +quickfix          +wildignore
+cscope            +listcmds          +reltime           +wildmenu
+cursorbind        +localmap          +rightleft         +windows
+cursorshape       -lua               -ruby              +writebackup
+dialog_con        +menu              +scrollbind        -X11
+diff              +mksession         +signs             -xfontset
+digraphs          +modify_fname      +smartindent       -xim
-dnd               +mouse             +startuptime       -xpm
-ebcdic            -mouseshape        +statusline        -xsmp
+emacs_tags        +mouse_dec         -sun_workshop      -xterm_clipboard
+eval              -mouse_gpm         +syntax            -xterm_save
+ex_extra          -mouse_jsbterm     +tag_binary
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: gcc -std=gnu99 -c -I. -Iproto -DHAVE_CONFIG_H     -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc -std=gnu99   -L/usr/local/lib -Wl,--as-needed -o vim        -lm -ltinfo -lnsl  -lselinux  -ldl     -L/usr/local/lib/python3.7/config-3.7m-x86_64-linux-gnu -lpython3.7m -lpthread -ldl -lutil -lm
```

The error is also reproducible using the non-stripped version of vim.
However, the symbol in question **is defined** in the non-stripped version (I'm assuming it also exists in the stripped/installed version - just hard to confirm with no debugging symbols):
```
$ nm --defined-only vim | grep PyByteArray_Type
0000000000aeac00 D PyByteArray_Type
```

It does **not occur** when running the installed (system) Python 3 interpreter.
As expected, the same symbol is also defined in the system interpreter:
```
$ nm --defined-only /usr/local/bin/python3 | grep PyByteArray_Type
00000000008f7140 D PyByteArray_Type
```

The vim interpreter was linked with the same Python 3 library as used by the system interpreter.

**Comment by tonymec** [comment]
1. What answer do you get to `python3 --version` at the shell prompt? (It should be 3.7 with possibly a third number.)
2. Do you get the same error if you type `:py3 import struct` after Vim has started?
3. If you get 3.7[.nn] at step 1 and the same error at step 2, try recompiling with `make reconfig`, which will rerun configure then recompile everything, so you should set your configure arguments in advance by means of environment variables, as explained at http://users.skynet.be/antoine.mechelynck/vim/compunix.htm (Caution: if you set them by means of a bash script, as I do, that script must be **sourced**, not just run, in the same shell where you will then compile.)

If after step 3 then the usual `make install` both run successfully you still get the same error with the newly compiled Vim, then I don't (yet?) know what to try next.

Best regards,
Tony.

**Comment by outeleiloa** [comment]
Thanks Tony.

1.  Python 3 version:
```
$ python3 --version
Python 3.7.1
```

2. Yes, the same exact error.

3. Unfortunately, no joy. Same problem

**Comment by outeleiloa** [comment]
Okay, so it _appears_ the issue can be resolved by pre-setting LDFLAGS with the following and then rebuilding vim:
```
$ export LDFLAGS="-rdynamic"
```

It appears to successfully resolve the symbol reference I was having issues with. 

Curiously, a brief look at ```if_python3.c``` implies the Python 3 library is treated as a shared library (which could be totally wrong as I haven't dug any deeper).
However, the actual library that's linked is static (although, this may be quite normal).

**Comment by outeleiloa** [comment]
For completeness, this is the current source I have checked out:
```
$ git log -n 1
commit 6b731886ca94d66b9bdedfb7e603af44a6400399
Author: Bram Moolenaar <Bram@vim.org>
Date:   Fri Nov 16 20:54:47 2018 +0100

    patch 8.1.0535: increment/decrement might get interrupted by updating folds

    Problem:    Increment/decrement might get interrupted by updating folds.
    Solution:   Disable fold updating for a moment. (Christian Brabandt,
                closes #3599)
```

**Comment by tonymec** [comment]
Bram's current source is actually 8.1.0541 as of this writing, but the latest patchlevel is Windows-only AFAICT so I'm using 8.1.0540 for the time being. I don't see anything related to solving your problem in patches 536 to 541 however.

My Huge build of (g)vim (the only one of my builds which has interpreters, not python3 however: it is built with `+lua -mzscheme +perl +python -python3 +ruby +tcl`) gets linked with `-rdynamic -Wl,-export-dynamic` near the start of its link command-line without the need for me to specify them explicitly.

@brammool @chrisbra : maybe `auto/configure` forgets to include these switches when building with python3 and no other interpreters? Or is it something else?

**Comment by outeleiloa** [comment]
So, the issue I've encountered could well be environmental.

Python 3 (3.7.1) was built [from source.](https://www.python.org/downloads/source/)

The OS is a relatively fresh install of CentOS:
```
$ cat /etc/*release
CentOS Linux release 7.5.1804 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

I'm not sure if anyone else has encountered a similar problem, so thinking of closing this as an environmental issue (unless someone can prove it's an actual bug in the vim source).

**Comment by outeleiloa** [comment]
Closing as environmental issue.

Simple workaround is [mentioned above](https://github.com/vim/vim/issues/3629#issuecomment-440845680).

**Comment by rdbisme** [comment]
I was hitting the same problem, me too on CentOs 7. Are we sure it's not a bug in the build scripts?

Workaround fixed also for me.

**Comment by maraal** [comment]
Same problem here, but inside a virtualenv. The workaround fixed only outside virtualenvs.

**Comment by Yggdroot** [comment]
Same problem here when compiling on Manjaro kde.
$ git log -1
```
commit 38854b565acba39eff36cf3c6396c911bf072bdc (HEAD -> master, tag: v9.0.0962, origin/master, origin/HEAD)
Author: porygonisaduck <alvaradx@umich.edu>
Date:   Sun Nov 27 20:55:05 2022 +0000

patch 9.0.0962: virtual text below cannot be placed below empty lines

Problem:    Virtual text below cannot be placed below empty lines.
Solution:   Add one character. (James Alvarado, closes #11606, closes #11520)

```
The workaround helps.
My command is:
```sh
LDFLAGS="-rdynamic" ./configure --prefix=/home/yggdroot/.vim9 --enable-luainterp  --enable-python3interp --enable-cscope --enable-gui
```
